File: src/main/java/com/pebble/baseAuth/persistence/UserRepositoryImpl.java
package com.pebble.baseAuth.persistence;

import com.pebble.baseAuth.domain.User;
import com.pebble.baseAuth.domain.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
@RequiredArgsConstructor
public class UserRepositoryImpl implements UserRepository {

    private final UserJpaRepository userJpaRepository;

    @Override
    public boolean existsByUsernameAndDeletedAtIsNull(String username) {
        return userJpaRepository.existsByUsernameAndDeletedAtIsNull(username);
    }

    @Override
    public Optional<User> findByUsernameAndDeletedAtIsNull(String username) {
        return userJpaRepository.findByUsernameAndDeletedAtIsNull(username)
                .map(UserEntity::toDomain);
    }

    @Override
    public Optional<User> findByProviderAndProviderIdAndDeletedAtIsNull(String provider, String providerId) {
        return userJpaRepository.findByProviderAndProviderIdAndDeletedAtIsNull(provider, providerId)
                .map(UserEntity::toDomain);
    }

    @Override
    public User save(User user) {
        UserEntity entity = UserEntity.from(user);
        return userJpaRepository.save(entity).toDomain();
    }
}

File: src/main/kotlin/com/pebble/baseAuth/persistence/UserRepositoryImpl.kt
package com.pebble.baseAuth.persistence

import com.pebble.baseAuth.domain.User
import com.pebble.baseAuth.domain.UserRepository
import org.springframework.stereotype.Repository
import java.util.Optional

@Repository
class UserRepositoryImpl(private val userJpaRepository: UserJpaRepository) : UserRepository {

    override fun existsByUsernameAndDeletedAtIsNull(username: String): Boolean {
        return userJpaRepository.existsByUsernameAndDeletedAtIsNull(username)
    }

    override fun findByUsernameAndDeletedAtIsNull(username: String): Optional<User> {
        return userJpaRepository.findByUsernameAndDeletedAtIsNull(username)
            .map { it.toDomain() }
    }

    override fun findByProviderAndProviderIdAndDeletedAtIsNull(provider: String, providerId: String): Optional<User> {
        return userJpaRepository.findByProviderAndProviderIdAndDeletedAtIsNull(provider, providerId)
            .map { it.toDomain() }
    }

    override fun save(user: User): User {
        val entity = UserEntity.from(user)
        return userJpaRepository.save(entity).toDomain()
    }
}
